apollo-router/src/plugins/file_uploads/mod.rs

```diff
@@ -0,0 +1,726 @@
+                for (first_file, (node, last_file)) in sequence.into_iter() {
+                    // FIXME: error
+                    assert!(first_file > sequence_last_file);
+                    sequence_last_file = last_file;
+                    nodes.push(node);
+                }
+
+                parallel.push(PlanNode::Sequence { nodes });
+            }
+
+            (files, PlanNode::Parallel { nodes: parallel })
+        }
+    }
+}
+
+async fn call_subgraph(mut req: subgraph::Request) -> subgraph::Request {
+    let supergraph_result = req
+        .context
+        .extensions()
+        .lock()
+        .get::<SupergraphLayerResult>()
+        .cloned();
+    if let Some(supergraph_result) = supergraph_result {
+        let SupergraphLayerResult {
+            multipart,
+            map_per_variable,
+            files_order,
+        } = supergraph_result;
+
+        let variables = &mut req.subgraph_request.body_mut().variables;
+        let mut map_field: MapField = IndexMap::new();
+        for (variable_name, variable_value) in variables.iter_mut() {
+            let variable_name = variable_name.as_str();
+            if let Some(variable_map) = map_per_variable.get(variable_name) {
+                for (file, paths) in variable_map.iter() {
+                    map_field.insert(
+                        file.clone(),
+                        paths
+                            .iter()
+                            .map(|path| {
+                                if path.is_empty() {
+                                    format!("variables.{}", variable_name)
+                                } else {
+                                    format!("variables.{}.{}", variable_name, path.join("."))
+                                }
+                            })
+                            .collect(),
+                    );
+                    for path in paths {
+                        if let Some(json_value) = try_path(variable_value, path) {
+                            json_value.take();
+                        }
+                    }
+                }
+            }
+        }
+        if !map_field.is_empty() {
+            map_field.sort_by_cached_key(|file, _| files_order.get_index_of(file));
+            req.subgraph_request
+                .extensions_mut()
+                .insert(SubgraphHttpRequestExtensions {
+                    multipart,
+                    map_field,
+                });
+        }
+    }
+    req
+}
+
+struct SubgraphHttpRequestExtensions {
+    multipart: Arc<Mutex<MultipartFileStream>>,
+    map_field: MapField,
+}
+
+const APOLLO_REQUIRE_PREFLIGHT: http::HeaderName =
+    HeaderName::from_static("apollo-require-preflight");
+const TRUE: http::HeaderValue = HeaderValue::from_static("true");
+
+pub(crate) async fn http_request_wrapper(
+    mut req: http::Request<hyper::Body>,
+) -> http::Request<hyper::Body> {
+    let supergraph_result = req.extensions_mut().remove();
```

Comment:
what happens if a request is retried? 2 examples:
- the request fails, but request retry was set up in traffic shaping
- using APQ to the subgraph, a first call is made with the hash, the subgraph rejects the request, the router retries with the query to register it

Reply:
I changed code to clone here in 1a852be922514b8f54e4e2d0227c785e5f1c4ab9
But I think we have a deeper problem here; if the subgraph rejects the request after some of the files were transferred, we can't replicate this file in retry.
I think we need to disable APQ for requests using file uploads.

Reply:
++, I know we planned on documenting it and calling it out, not sure where we are atm but if anything it will need to be a followup.